libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
         return requestBodyString.toRequestBody(mediaType)
     }
 
+    // region Salesforce Identity API Headless Registration Flow
+
+    override suspend fun startRegistration(
+        email: String,
+        firstName: String,
+        lastName: String,
+        username: String,
+        newPassword: String,
+        reCaptchaToken: String,
+        otpVerificationMethod: OtpVerificationMethod
+    ): StartRegistrationResult {
+        // Validate parameters.
+        val trimmedEmail = email.trim()
+        if (!EMAIL_ADDRESS.matcher(trimmedEmail).matches()) {
+            return StartRegistrationResult(InvalidEmail)
+        }
+        val trimmedNewPassword = newPassword.trim()
+        trimmedNewPassword.mapInvalidPasswordToResult()?.let { return StartRegistrationResult(it) }
+        val trimmedUsername = username.trim()
+        trimmedUsername.mapInvalidUsernameToResult()?.let { return StartRegistrationResult(it) }
+        val reCaptchaParameterGenerationResult = generateReCaptchaParameters(reCaptchaToken)
+
+        // Determine the OTP verification method.
+        val otpVerificationMethodString = otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue
+        // Generate the start registration request body.
+        val startRegistrationRequestBodyString = runCatching {
+            StartRegistrationRequestBody(
+                recaptcha = reCaptchaParameterGenerationResult.nonEnterpriseReCaptchaToken,
+                recaptchaevent = reCaptchaParameterGenerationResult.enterpriseReCaptchaEvent,
+                userData = UserData(
+                    email = trimmedEmail,
+                    username = trimmedUsername,
+                    password = trimmedNewPassword,
+                    firstName = firstName.trim(),
+                    lastName = lastName.trim()
+                ),
+                otpVerificationMethod = otpVerificationMethodString
+            ).toJson()
+        }.onFailure { e ->
+            SalesforceSDKLogger.e(TAG, "Cannot JSON encode start registration request body due to an encoding error with message '${e.message}'.", e)
+            return StartRegistrationResult(UnknownError)
+        }.getOrNull()
+
+        // Create the start registration request.
+        val startRegistrationRequest = RestRequest(
+            POST,
+            LOGIN,
+            "$loginUrl$OAUTH_ENDPOINT_HEADLESS_INIT_REGISTRATION",
```

Comment:
NIT:  I know I have said this before, but I truly think this is unreadable.  I don't think this is what the designers or doc writers had in mind when said to prefer string templates over concatenation.  Ever single example they give has string literals and spaces around the variables used.  You wouldn't pass `"$loginUrl"` as a parameter, right?